stdlib/array.gr

```diff
@@ -985,3 +985,49 @@ provide let rotate = (n, arr) => {
     }
   }
 }
+
+let rec _chunk = (chunkSize, allowPartial, arr, arrLen, chunkStartIndex) => {
+  if (arrLen == 0 || arrLen < chunkSize && !allowPartial) {
+    []
+  } else if (arrLen < chunkSize) {
+    [slice(chunkStartIndex, length(arr), arr)]
+  } else {
+    // create the first chunk of the given array
+    let firstChunk = slice(chunkStartIndex, chunkStartIndex + chunkSize, arr)
+    let restChunks = _chunk(
+      chunkSize,
+      allowPartial,
+      arr,
+      arrLen - chunkSize,
+      chunkStartIndex + chunkSize
+    )
+    [firstChunk, ...restChunks]
+  }
+}
+
+/**
+ * Splits the given array into chunks of the provided size.
+ *
+ * @param chunkSize: The size of each chunk
+ * @param allowPartial: Whether to allow partial chunks at the end of the array
+ * @param arr: The array to chunk
+ * @returns An array of chunks
+ *
+ * @throws InvalidArgument(String): When `chunkSize` is not an integer
+ * @throws InvalidArgument(String): When `chunkSize` is negative
+ *
+ * @example chunk(2, false, [> 1, 2, 3, 4]) == [> [> 1, 2], [> 3, 4] ]
+ * @example chunk(2, false, [> 1, 2, 3, 4, 5]) == [> [> 1, 2], [> 3, 4] ]
+ * @example chunk(2, true, [> 1, 2, 3, 4, 5]) == [> [> 1, 2], [> 3, 4], [> 5] ]
+ * @since v0.6.0
+ */
+provide let chunk = (chunkSize, allowPartial, arr) => {
+  if (chunkSize == 0) {
+    [>]
```

Comment:
Should this case throw an exception?

Reply:
I think that having it return an empty array is fine behaviour certainly subjective though. I prefer not to throw exception where possible though and I think that having it return an empty array is behaviour that makes sense.

Reply:
Having a chunk size of 0 doesn't really make sense to me though, in the same way as having a negative chunk size. On a related note calling the `windows` function on slices in Rust (basically the equivalent of the function you implemented) with an argument of 0 results in a panic.

Reply:
That's fair enough. I can implement an exception there then.